lib/mix/lib/mix/tasks/format.ex

```diff
@@ -390,10 +482,40 @@ defmodule Mix.Tasks.Format do
     end
   end
 
-  defp find_formatter_opts_for_file(split, {formatter_opts, subs}) do
+  defp find_formatter_for_file(file, formatter_opts) do
+    ext = Path.extname(file)
+
+    cond do
+      ext in ~w(.ex .exs) ->
+        &elixir_format(&1, [file: file] ++ formatter_opts)
+
+      plugin = find_plugin_for_extension(formatter_opts, ext) ->
+        &plugin.format(&1, formatter_opts)
+
+      true ->
+        & &1
+    end
+  end
+
+  defp find_plugin_for_extension(formatter_opts, ext) do
+    plugins = Keyword.get(formatter_opts, :plugins, [])
+
+    Enum.find(plugins, fn plugin ->
+      Code.ensure_loaded?(plugin) and function_exported?(plugin, :features, 1) and
```

Comment:
Is this required? It seems the list of plugins is already validated in `eval_deps_and_subdirectories`.

Reply:
They are validated BUT they are still kept because we only validate it when .formatter.exs is loaded but then the result of .formatter.exs is cached. So we need to look at the cached result and see if they have been loaded post caching or not.